### PR TITLE
Add Koji handler for build complete and build tag/untag messages

### PR DESCRIPTION
**Pull request by sarah256**

### CODE REVIEW

File: estuary_updater/handlers/base.py [hunk 1]
```diff
@@ -89,42 +89,52 @@ def is_container_build(self, build_info):
...
             return False
 
-    def get_or_create_build(self, identifier, original_nvr=None, force_container_label=False):
+    def get_or_create_build(self, identifier, build_info=None, original_nvr=None,
+                            force_container_label=False):
         """
         Get a Koji build from Neo4j, or create it if it does not exist in Neo4j.
 
         :param str/int identifier: an NVR (str) or build ID (int)
+        :param dict build_info: all info needed to create a build, else get from a Koji session
```
Comment (mprahl) [inline estuary_updater/handlers/base.py]: This is actually a `kwarg` and not a `param`.

I'd also replace `else get from a Koji session` with `else get it from Koji`

File: estuary_updater/handlers/base.py [hunk 2]
```diff
@@ -89,42 +89,52 @@ def is_container_build(self, build_info):
...
-            'creation_time': datetime.fromtimestamp(int(koji_build_info['creation_ts'])),
-            'epoch': koji_build_info['epoch'],
-            'extra': json.dumps(koji_build_info['extra']),
-            'id_': str(koji_build_info['id']),
-            'name': koji_build_info['package_name'],
-            'release': koji_build_info['release'],
-            'start_time': datetime.fromtimestamp(int(koji_build_info['start_ts'])),
-            'state': koji_build_info['state'],
-            'version': koji_build_info['version']
+            'completion_time': None,
```
Comment (mprahl) [inline estuary_updater/handlers/base.py]: Instead of making this `None`, you could just delete these line since you set it later on. Same with the other two.

File: estuary_updater/handlers/koji.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        Handle a message and update Neo4j if necessary.
+
+        :param dict msg: a message to be processed
+        """
+        topic = msg['topic']
+
+        if topic == '/topic/VirtualTopic.eng.brew.build.complete':
+            self.build_complete_handler(msg)
+        elif topic == '/topic/VirtualTopic.eng.brew.build.tag' or \
```
Comment (mprahl) [inline estuary_updater/handlers/koji.py]: This will always evaluate to `True`. You want to do:
```python
elif topic == '/topic/VirtualTopic.eng.brew.build.tag' or topic == '/topic/VirtualTopic.eng.brew.build.untag':
```

File: estuary_updater/handlers/koji.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            'completion_ts': msg['body']['msg']['info']['completion_ts'],
+            'extra': msg['body']['msg']['info']['extra'],
+            'package_name': msg['body']['msg']['info']['name'],
+            'version': msg['body']['msg']['info']['version'],
+            'release': msg['body']['msg']['info']['release'],
+            'owner_name': msg['body']['msg']['info']['owner_name']
+        }
+
+        self.get_or_create_build(msg['body']['msg']['info']['id'], build_info=build_info,
+                                 original_nvr=msg['body']['msg']['info']['nvr'])
```
Comment (mprahl) [inline estuary_updater/handlers/koji.py]: `original_nvr` is only a value for container builds built by Freshmaker. Drop that from the method call and it should be good afterwards.

File: estuary_updater/handlers/koji.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+        :param dict msg: a message to be processed
+        """
+        tag = KojiTag.create_or_update({
+            'id_': msg['body']['msg']['tag']['id'],
+            'name': msg['body']['msg']['tag']['name']
+        })[0]
+
+        try:
+            build = KojiBuild.get_or_create({'id_': msg['body']['msg']['build']['id']})[0]
```
Comment (mprahl) [inline estuary_updater/handlers/koji.py]: This can be outside of the try except.

File: tests/handlers/test_koji.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    assert build.version == '17.4.0'
+    assert build.start_time == datetime(2018, 8, 3, 17, 49, 42, tzinfo=pytz.utc)
+    assert build.state == 0
+
+
+def test_build_tag():
+    """Test the Koji handler when it recieves a new build tag message."""
+    with open(path.join(message_dir, 'koji', 'build_tag.json'), 'r') as f:
+        msg = json.load(f)
+    assert KojiHandler.can_handle(msg) is True
```
Comment (mprahl) [inline tests/handlers/test_koji.py]: The build should exist before the tag so can you create a dummy build? You could reuse the pytest fixture we used in the Errata Tool and Freshmaker scraper.

File: tests/handlers/test_koji.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    tag = KojiTag.nodes.get_or_none(id_=16413)
+
+    assert tag.builds.is_connected(build)
+
+
+def test_build_untag():
+    """Test the Koji handler when it recieves a new build untag message."""
+    with open(path.join(message_dir, 'koji', 'build_untag.json'), 'r') as f:
+        msg = json.load(f)
+    assert KojiHandler.can_handle(msg) is True
```
Comment (mprahl) [inline tests/handlers/test_koji.py]: The tag and build should exist in Neo4j and the build should be tagged to it before this handler runs. Could you simulate that?

File: tests/messages/koji/build_complete.json [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            "content-length": "864", 
+            "expires": "0", 
+            "old": "BUILDING", 
+            "JMS_AMQP_MESSAGE_FORMAT": "0", 
+            "timestamp": "0", 
+            "JMS_AMQP_NATIVE": "false", 
+            "destination": "/topic/VirtualTopic.eng.brew.build.complete", 
+            "priority": "4", 
+            "subscription": "/queue/Consumer.client-datanommer.upshift-prod.VirtualTopic.eng.>", 
+            "message-id": "ID:messaging-devops-broker01.web.prod.ext.phx2.redhat.com-41951-1532471159894-4:793878:0:0:1",
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: Obscuring this would be nice.

File: tests/messages/koji/build_complete.json [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                    "original_url": "git://pkgs.devel.redhat.com/rpms/python-attrs?#3be3cb33e6432d8392ac3d9e6edffd990f618432"
+                  }
+                }, 
+                "creation_time": "2018-08-03 17:49:42.735506", 
+                "completion_time": null, 
+                "package_id": 63243, 
+                "id": 736244, 
+                "build_id": 736244, 
+                "state": 0, 
+                "source": "git://pkgs.devel.redhat.com/rpms/python-attrs#3be3cb33e6432d8392ac3d9e6edffd990f618432",
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: Obscuring this would be nice.

File: tests/messages/koji/build_complete.json [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            "JMS_AMQP_FirstAcquirer": "false", 
+            "type": "BuildStateChange", 
+            "name": "python-attrs"
+        },
+        "msg":{
+            "info": {
+                "package_name": "python-attrs", 
+                "extra": {
+                  "source": {
+                    "original_url": "git://pkgs.devel.redhat.com/rpms/python-attrs?#3be3cb33e6432d8392ac3d9e6edffd990f618432"
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: Obscuring this would be nice.

File: tests/messages/koji/build_complete.json [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        "content-length": "864", 
+        "expires": "0", 
+        "old": "BUILDING", 
+        "JMS_AMQP_MESSAGE_FORMAT": "0", 
+        "timestamp": "0", 
+        "JMS_AMQP_NATIVE": "false", 
+        "destination": "/topic/VirtualTopic.eng.brew.build.complete", 
+        "priority": "4", 
+        "subscription": "/queue/Consumer.client-datanommer.upshift-prod.VirtualTopic.eng.>", 
+        "message-id": "ID:messaging-devops-broker01.web.prod.ext.phx2.redhat.com-41951-1532471159894-4:793878:0:0:1",
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: Obscuring this would be nice.

File: tests/messages/koji/build_tag.json [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+                "volume_id": 0, 
+                "creation_ts": 1533308980.75512, 
+                "name": "module-build-macros", 
+                "task_id": 17577398, 
+                "volume_name": "DEFAULT", 
+                "release": "1.el8+1318+37d5dc6c"
+              }, 
+              "user": {
+                "status": 0, 
+                "krb_principal": "mbs/mbs-backend.host.prod.eng.bos.redhat.com@REDHAT.COM",
```
Comment (mprahl) [inline tests/messages/koji/build_tag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        "headers":{
+            "content-length": "1273", 
+            "destination": "/topic/VirtualTopic.eng.brew.build.untag", 
+            "JMS_AMQP_MESSAGE_FORMAT": "0", 
+            "JMS_AMQP_NATIVE": "false", 
+            "expires": "0", 
+            "tag": "rhos-13.0-rhel-7-pending", 
+            "priority": "4", 
+            "subscription": "/queue/Consumer.client-datanommer.upshift-prod.VirtualTopic.eng.>", 
+            "message-id": "ID:messaging-devops-broker01.web.prod.ext.phx2.redhat.com-44167-1533567716468-4:3791:0:0:1",
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+              "arches": null, 
+              "maven_include_all": false, 
+              "id": 15638
+            }, 
+            "force": false, 
+            "build": {
+              "package_name": "openstack-tripleo-heat-templates", 
+              "extra": {
+                "source": {
+                  "original_url": "git://pkgs.devel.redhat.com/rpms/openstack-tripleo-heat-templates?#3fcaf715e349336c0b2062fe7b1fa4369c12852f"
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+                }
+              }, 
+              "creation_time": "2018-08-06 08:26:22.378117", 
+              "completion_time": "2018-08-06 08:30:13.865342", 
+              "package_id": 48664, 
+              "id": 739457, 
+              "build_id": 739457, 
+              "state": 1, 
+              "source": "git://pkgs.devel.redhat.com/rpms/openstack-tripleo-heat-templates#3fcaf715e349336c0b2062fe7b1fa4369c12852f",
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+              "volume_id": 0, 
+              "creation_ts": 1533543982.37812, 
+              "name": "openstack-tripleo-heat-templates", 
+              "task_id": 17620334, 
+              "volume_name": "DEFAULT", 
+              "release": "11.el7ost"
+            }, 
+            "user": {
+              "status": 0, 
+              "krb_principal": "rcm/build@REDHAT.COM",
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    "headers":{
+        "content-length": "1273", 
+        "destination": "/topic/VirtualTopic.eng.brew.build.untag", 
+        "JMS_AMQP_MESSAGE_FORMAT": "0", 
+        "JMS_AMQP_NATIVE": "false", 
+        "expires": "0", 
+        "tag": "rhos-13.0-rhel-7-pending", 
+        "priority": "4", 
+        "subscription": "/queue/Consumer.client-datanommer.upshift-prod.VirtualTopic.eng.>", 
+        "message-id": "ID:messaging-devops-broker01.web.prod.ext.phx2.redhat.com-44167-1533567716468-4:3791:0:0:1",
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: tests/messages/koji/build_untag.json [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        "content-length": "1273", 
+        "destination": "/topic/VirtualTopic.eng.brew.build.untag", 
+        "JMS_AMQP_MESSAGE_FORMAT": "0", 
+        "JMS_AMQP_NATIVE": "false", 
+        "expires": "0", 
+        "tag": "rhos-13.0-rhel-7-pending", 
+        "priority": "4", 
+        "subscription": "/queue/Consumer.client-datanommer.upshift-prod.VirtualTopic.eng.>", 
+        "message-id": "ID:messaging-devops-broker01.web.prod.ext.phx2.redhat.com-44167-1533567716468-4:3791:0:0:1", 
+        "user": "rcm/build",
```
Comment (mprahl) [inline tests/messages/koji/build_untag.json]: Obscuring this would be nice.

File: estuary_updater/handlers/base.py [hunk 19]
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: GPL-3.0+
 
-from __future__ import unicode_literals
+from __future__ import unicode_literals, absolute_import
```
Comment (mprahl) [inline estuary_updater/handlers/base.py]: For consistency in the project, could you file a separate PR with this import added to all of our files? That way one file doesn't import differently than the rest.

File: tests/messages/koji/build_complete.json [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+                  "source": {
+                    "original_url": "git://pkgs.domain.com/rpms/python-attrs?#3be3cb33e6432d8392ac3d9e6edffd990f618432"
+                  }
+                }, 
+                "creation_time": "2018-08-03 17:49:42.735506", 
+                "completion_time": null, 
+                "package_id": 63243, 
+                "id": 736244, 
+                "build_id": 736244, 
+                "state": 0,
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: This is the value you're keying on but unfortunately it is incorrect. The state should be 1 to denote it is complete. You could use the state in the "new" key below for the correct value or hard-code it.

File: tests/messages/koji/build_complete.json [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        "msg":{
+            "info": {
+                "package_name": "python-attrs", 
+                "extra": {
+                  "source": {
+                    "original_url": "git://pkgs.domain.com/rpms/python-attrs?#3be3cb33e6432d8392ac3d9e6edffd990f618432"
+                  }
+                }, 
+                "creation_time": "2018-08-03 17:49:42.735506", 
+                "completion_time": null,
```
Comment (mprahl) [inline tests/messages/koji/build_complete.json]: This is curious... Unfortunately we need this value but it's not set in the message. I filed:
https://pagure.io/koji/issue/1033

**Comment by mprahl** [comment]
@sarah256 it also looks like you forgot to create and connect the DistGitCommits that were used as the source for the build. See the scraper for an example:
https://github.com/release-engineering/estuary-api/blob/3fb8182bda7776167e31fbe21aff3d7d870322f1/scrapers/koji.py#L203

File: tests/messages/koji/build_complete.json (hunk 21 above)
Comment (sarah256) [reply tests/messages/koji/build_complete.json]: @mprahl actually may not have to refactor, but would like to discuss when we get the chance

**Comment by sarah256** [comment]
@mprahl changes were made and the commit connection was added -- im unsure about the best way to approach the missing bits of data from the messages but we can talk about that if you'd like, or if you think its fine how ive done it here thats fine too

File: estuary_updater/handlers/base.py [hunk 22]
```diff
@@ -89,42 +89,53 @@ def is_container_build(self, build_info):
...
             return False
 
-    def get_or_create_build(self, identifier, original_nvr=None, force_container_label=False):
+    def get_or_create_build(self, identifier, build_info=None, original_nvr=None,
+                            force_container_label=False):
         """
         Get a Koji build from Neo4j, or create it if it does not exist in Neo4j.
 
         :param str/int identifier: an NVR (str) or build ID (int)
+        :kwarg dict build_info: all info needed to create a build, else get it from Koji
```
Comment (mprahl) [inline estuary_updater/handlers/base.py]: I'd say since it's unlikely the messages will be fixed to include the proper build info in the near future, I'd say you can just remove this option and call the API every time. We will reimplement something like this when it does get fixed,

File: estuary_updater/handlers/freshmaker.py [hunk 23]
```diff
@@ -122,4 +122,4 @@ def create_or_update_build(self, build, event_id):
         # It's always going to be a container build when the build comes from Freshmaker, so we can
         # just set force_container_label to avoid unncessary heuristic checks
         return self.get_or_create_build(
-            koji_build_id, build['original_nvr'], force_container_label=True)
+            koji_build_id, original_nvr=build['original_nvr'], force_container_label=True)
```
Comment (mprahl) [inline estuary_updater/handlers/freshmaker.py]: This is an unnecessary change. This must be left over from your latest rebase.

File: estuary_updater/handlers/koji.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+    def build_complete_handler(self, msg):
+        """
+        Handle a build complete message and update Neo4j if necessary.
+
+        :param dict msg: a message to be processed
+        """
+        build = self.get_or_create_build(msg['body']['msg']['info']['id'])
+
+        commit_hash = msg['body']['msg']['info']['source'].split('#')[1]
```
Comment (mprahl) [inline estuary_updater/handlers/koji.py]: This will work in cases where the build is an RPM or container image, but this won't work for some other types. I'll send you a message showing this.

Comment (mprahl) [reply estuary_updater/handlers/koji.py]: My suggestion would be to ignore non-RPM and container image builds like we do in the scraper but we'll have to find a way to do that (likely going to have to look at the build task).

File: tests/handlers/test_koji.py [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+from tests import message_dir
+from estuary_updater.handlers.koji import KojiHandler
+from estuary_updater import config
+
+
+@mock.patch('koji.ClientSession')
+def test_build_complete(mock_koji_cs, mock_getBuild_two):
+    """Test the Koji handler when it recieves a new build complete message."""
+    mock_koji_session = mock.Mock()
+    mock_getBuild_two['state'] = 1
```
Comment (mprahl) [inline tests/handlers/test_koji.py]: This should actually be in the building state in Neo4j and then your handler marks it as complete.
You can do that with `koji.BUILD_STATES['BUILDING']`

File: tests/handlers/test_koji.py [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+    assert build.extra == ('{"source": {"original_url": "git://pkgs.domain.com/rpms/python-'
+                           'attrs?#3be3cb33e6432d8392ac3d9e6edffd990f618432"}}')
+    assert build.id_ == '736244'
+    assert build.name == 'python-attrs'
+    assert build.release == '8.el8+1325+72a36e76'
+    assert build.version == '17.4.0'
+    assert build.start_time == datetime(2018, 8, 3, 17, 49, 42, tzinfo=pytz.utc)
+    assert build.state == 1
+
+    commit = DistGitCommit.get_or_create({
```
Comment (mprahl) [inline tests/handlers/test_koji.py]: This should be `DistGitCommit.nodes.get_or_none(hash_='3be3cb33e6432d8392ac3d9e6edffd990f618432')` since you don't want to create it in the test if doesn't exist.

File: tests/handlers/test_koji.py (hunk 25 above)
Comment (mprahl) [reply tests/handlers/test_koji.py]: I made a mistake. I was thinking this was a build in Neo4j and not getBuild mock output.

File: estuary_updater/handlers/koji.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+    def build_complete_handler(self, msg):
+        """
+        Handle a build complete message and update Neo4j if necessary.
+
+        :param dict msg: a message to be processed
+        """
+        commit_hash_pattern = re.compile(r'(?:\#)([0-9a-f]{40})$')
+        commit_hash = re.findall(commit_hash_pattern, msg['body']['msg']['info']['source'])
+
+        if commit_hash:
```
Comment (mprahl) [inline estuary_updater/handlers/koji.py]: Please add a comment explaining this as discussed earlier.
